File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0053_Maximum_Subarray.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0053_Maximum_Subarray {
    public int maxSubArray(int[] nums) {
        int res = Integer.MIN_VALUE;
        if (nums == null || nums.length == 0) {
            return res;
        }
        int cur = 0;
        for (int n : nums) {
            cur += n;
            res = Math.max(res, cur);
            if (cur < 0) {
                cur = 0;
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0053_Maximum_Subarray.cc
#include <limits.h>
#include <vector>

using namespace std;

class Solution {
 public:
  int maxSubArray(vector<int>& nums) {
    if (nums.size() == 0) {
      return 0;
    }
    int res = INT_MIN, cur = 0;
    for (int i = 0; i < nums.size(); i++) {
      cur += nums[i];
      res = max(res, cur);
      if (cur < 0) {
        cur = 0;
      }
    }
    return res;
  }
};
